src/role.ts

```diff
@@ -0,0 +1,380 @@
+      .filter((x) => typeof x === 'string'),
+    ) as string[];
+    const resourceNames = normalizedArray(resources
+      .map((resource) => resource.name)
+      .filter((x) => typeof x === 'string'),
+    ) as string[];
+
+    this.addRule({
+      apiGroups,
+      resources: resourceTypes,
+      resourceNames,
+      verbs,
+    });
+  }
+
+  /**
+   * List of rules included in this rule.
+   *
+   * Returns a copy. To add a rule, use `addRule()`.
+   */
+  public get rules() {
+    return [...this._rules];
+  }
+}
+
+/**
+ * A reference to any Role or ClusterRole.
+ */
+export interface IRole extends IResource {
+
+}
+
+/**
+ * Options for `Role`.
+ */
+export interface RoleProps extends RoleCommonProps {
+  /**
+   * The namespace of the resources this role's permissions should apply to
+   * (required).
+   */
+  readonly namespace: string;
+}
+
+/**
+ * Role is a namespaced, logical grouping of PolicyRules that can be referenced
+ * as a unit by a RoleBinding.
+ */
+export class Role extends RoleBase implements IRole {
+  /**
+   * @see base.Resource.apiObject
+   */
+  protected readonly apiObject: ApiObject;
+
+  public readonly resourceType = 'roles';
+
+  constructor(scope: Construct, id: string, props: RoleProps) {
+    super(scope, id);
+
+    if (props.metadata?.namespace && props.metadata.namespace !== props.namespace) {
+      throw new Error('If `metadata.namespace` is passed as an option, its value must match the value of `namespace`.');
+    }
+
+    const metadata: ApiObjectMetadata = {
+      ...props.metadata,
+      namespace: props.namespace,
+    };
+
+    this.apiObject = new k8s.KubeRole(this, 'Resource', {
+      metadata,
+      rules: Lazy.any({ produce: () => this.synthesizeRules() }),
+    });
+  }
+
+  private synthesizeRules(): any {
+    return undefinedIfEmpty(this._rules.map(rule => rule._toKube()));
+  }
+
+  /**
+   * Create a RoleBinding that binds the permissions in this Role
+   * to a list of subjects, that will only apply this role's namespace.
+   * @param subjects a list of subjects to bind to
+   */
+  public bind(...subjects: ISubject[]): RoleBinding {
+    const binding = new RoleBinding(this, 'RoleBinding', {
+      namespace: this.metadata.namespace!,
+      role: this,
+    });
+    binding.addSubjects(...subjects);
+    return binding;
+  }
+}
+
+/**
+ * Represents a cluster-level role.
+ */
+export interface IClusterRole extends IResource {
+
+}
+
+/**
+ * Options for `ClusterRole`.
+ */
+export interface ClusterRoleProps extends RoleCommonProps {
+  /**
+   * Specify labels that should be used to locate ClusterRoles, whose rules
+   * will be automatically filled into this ClusterRole's rules.
+   */
+  readonly aggregationLabels?: { [key: string]: string };
+}
+
+/**
+ * ClusterRole is a cluster level, logical grouping of PolicyRules that can be
+ * referenced as a unit by a RoleBinding or ClusterRoleBinding.
+ */
+export class ClusterRole extends RoleBase implements IClusterRole {
```

Comment:
I'm probably going to skip this since to get some inheritance behavior working I need `Role` and `ClusterRole` to both have access to the un-exported `RoleBase` class (otherwise JSII throws errors preventing me from giving `Role.addRule(rule: ResourcePolicyRuleProps)` a different signature from `RoleBase.addRule(rule: PolicyRuleProps)`).